optax/contrib/_muon.py

```diff
@@ -297,3 +297,151 @@ def muon(
           lambda x: 'muon' if x.ndim == 2 else 'adam', params
       ),
   )
+
+
+class MuonClipState(NamedTuple):
+  eta: list[jax.Array] | None = None
+
+
+def scale_by_clip_muon() -> base.GradientTransformationExtraArgs:
+  """
+  Rescale ("clip") the weight after gradient updates by a factor of eta_bar from
+  a factor of eta.
+
+  Optax uses additive updates and the weight is assumed stored scale, the update
+  is:
+
+  g = d(eta * W)  # gradient of the scaled parameter
+  eta * W + updates = eta_bar * W_bar = eta_bar * (W + d(W))
+
+  updates = (eta_bar - eta) * W + eta_bar * d(W)
+
+  updates = (eta_bar - eta) * W + eta_bar / eta * d(eta * W)
+  """
+
+  def init_fn(params):
+    del params
+    return MuonClipState()
+
+  def update_fn(updates, state, params=None, eta_bar=None):
+    assert eta_bar is not None
+    flat_updates, update_struct = jax.tree.flatten(updates)
+    flat_params = jax.tree.leaves(params)
+    if len(flat_updates) != len(flat_params) != len(eta_bar):
+      raise ValueError("In MuonClip, the length of updates, params, eta_bar and"
+                       f" eta must be equal, but got {len(flat_updates)=},"
+                       f" {len(flat_params)=}, {len(eta_bar)=}.")
+    if state.eta is None:
+      state.eta = eta_bar
+    if len(state.eta_bar) != len(eta_bar):
+      raise ValueError("In MuonClip, the length of eta_bar and eta must be"
+                       f" equal, but got {len(eta_bar)=}, {len(state.eta)=}.")
+    eps = jnp.finfo(flat_params[0]).eps if len(flat_params) else 1e-7
+
+    updates = [
+      (eta_bar - eta) * param + eta_bar / jnp.maximum(eta, eps) * update
+      for eta_bar, eta, param, update in zip(eta_bar, state.eta, flat_params,
+                                             flat_updates)]
+    return (
+      jax.tree.unflatten(updates, update_struct), MuonClipState(eta=eta_bar)
+    )
+
+  return base.GradientTransformationExtraArgs(init_fn, update_fn)
+
+
+def _example_qk_proj_param_label_fn(params: base.Params):
+    params_with_paths, treedef = jax.tree.flatten_with_path(params)
+
+    def mask_fn(path_param):
+        path, param = path_param
+        str_path = "/".join(map(str, path))
+        if param.ndim == 2 and "q_proj" in str_path:
```

Comment:
shouldn't you also use moun_clip for "k_proj" ?

Reply:
oh, right, definitely